vdr-plugin-epgsearch/doc-src/en/createcats.1.txt
=head1 NAME

B<createcats> - helps you creating your own F<epgsearchcats.conf>

=head1 SYNOPSIS

B<createcats> [OPTIONS] F</path_to/epg.data>

=head1 DESCRIPTION

This tool is deliverd with the plugin and should exist in the plugins
source directory after compilation. It helps you in creating your own
epgsearchcats.conf, if the samples in directory 'conf' don't fit your needs.

createcats takes your epg.data as argument and scans it for suitable
EPG infos. Such an info is a set of a category name and a coresponding
value at the beginning of a line (represented with '|') and has the
form '|category: value', e.g.

 |Genre: Action

So simply call it with

B<createcats> /path_to/epg.data

The output is a file epgsearchcats.conf, that should be copied to your
plugins config dir. Before using it, you should do some customizing,
since not all things in the file will be suitable to be used as
extended EPG info.

=head1 OPTIONS

The full set of arguments is:

usage: B<createcats> [OPTIONS] F</path_to/epg.data>

 -m N, --minappearance=N    the minimum number a category has to appear
                            to be used
 -v N, --maxvalues=N        values of a category are omitted if they exceed
                            this number
 -l N, --maxlength=N        the maximum length of a text to be accepted
                            as a category value
 -h, --help                 this help

Some notes:

=over 4

=item -m N, --minappearance=N

createcats counts how often a category is used in your current
epg.data. If this is less than N, then this category will not be part
of the resulting epgsearchcats.conf.

=item -v N, --maxvalues=N

if the number of values for a category is more than N then createcats
will not output a value list for this category. As a consequence in
epgsearch the menu item gets an edit item, that can be filled with
every text. Else, the item gets a selection item, that lets you select
from a list of values.

=item -l N, --maxlength=N

if the text length of a value is more than N, this value is not part
of the value list.

=back

B<Hint:> Results are best, if your current EPG is as big as possible. So
update it, if you are using an external EPG provider, before calling
createcats.

Please edit this file now, since it will contain also things not suitable.

After that copy it to your plugins config directory.

=head1 SEE ALSO

C<epgsearch(1)>, C<epgsearch(4)>, C<epgsearchcats.conf(5)>

=head1 AUTHOR (man pages)

Mike Constabel <epgsearch (at) constabel (dot) net>

=head1 REPORT BUGS

Bugreports (german):

L<http://projects.vdr-developer.org/projects/plg-epgsearch>

Mailinglist:

L<http://www.vdr-developer.org/mailman/listinfo/epgsearch>


=head1 COPYRIGHT and LICENSE

Copyright (C) 2004-2010 Christian Wieninger

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
Or, point your browser to http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

The author can be reached at cwieninger@gmx.de

The project's page is at http://winni.vdr-developer.org/epgsearch

The MD5 code is derived from the RSA Data Security, Inc. MD5 Message-Digest Algorithm.


dbus/NEWS
dbus 1.13.18 (2020-07-02)
=========================

The “carnivorous border” release.

Maybe security fixes:

• On Unix, avoid a use-after-free if two usernames have the same
  numeric uid. In older versions this could lead to a crash (denial of
  service) or other undefined behaviour, possibly including incorrect
  authorization decisions if <policy group=...> is used.
  Like Unix filesystems, D-Bus' model of identity cannot distinguish
  between users of different names with the same numeric uid, so this
  configuration is not advisable on systems where D-Bus will be used.
  Thanks to Daniel Onaca.
  (dbus#305, dbus!166; Simon McVittie)

Other fixes:

• On Solaris and its derivatives, if a cmsg header is truncated, ensure
  that we do not overrun the buffer used for fd-passing, even if the
  kernel tells us to.
  (dbus#304, dbus!165; Andy Fiddaman)

• When built with CMake, use GNUInstallDirs' special-cases for prefixes
  /, /usr and /opt/*
  (dbus!155, Ralf Habacker)

• When built with CMake on Linux, allow systemd-specific features to be
  enabled, for feature parity with Autotools
  (dbus!155, Ralf Habacker)

• When built with CMake, install the same example files as with Autotools
  (dbus!155, Ralf Habacker)

• Correct the doc-comment for DBUS_ERROR_SPAWN_NO_MEMORY
  (dbus!163, Marc-André Lureau)

dbus 1.13.16 (2020-06-02)
=========================

The “ominous mushroom hat” release.

Denial of service fixes:

• CVE-2020-12049: If a message contains more file descriptors than can
  be sent, close those that did get through before reporting error.
  Previously, a local attacker could cause the system dbus-daemon (or
  another system service with its own DBusServer) to run out of file
  descriptors, by repeatedly connecting to the server and sending fds that
  would get leaked.
  Thanks to Kevin Backhouse of GitHub Security Lab.
  (dbus#294, GHSL-2020-057; Simon McVittie)

Enhancements:

• The API reference manual can be built as a Qt compiled help file if
  qhelpgenerator(-qt5) is available. This is controlled by
  --enable-qt-help and --with-qchdir in the Autotools build, or
  -DENABLE_QT_HELP and -DINSTALL_QCH_DIR in CMake.
  (dbus!150, Ralf Habacker)

Fixes:

• When built for Windows, return all autolaunch error information in
  the DBusError rather than printing some of it to stderr
  (dbus#191, dbus!131; Ralf Habacker)

• When built for Windows, don't truncate long log messages
  (dbus!134, Ralf Habacker)

• When built using CMake for a Unix platform, dbus-cleanup-sockets and
  dbus-uuidgen are now included (dbus!154, Ralf Habacker)

• When built for Windows with verbose mode enabled, don't print debugging
  messages related to poll() emulation into a fixed-size buffer that
  could overflow (dbus!125, Ralf Habacker)

• Adjust .desktop file parser to avoid a Coverity false positive
  (dbus!146, Coverity CID 354884; Ralf Habacker)

• Print shell-test diagnostics to stderr, avoiding warnings or errors
  from strict TAP parsers (dbus!157, Félix Piédallu)

Tests and CI enhancements:

• When the CI cross-builds Windows binaries on Linux, run unit tests
  using Wine (dbus#296, dbus!158; Ralf Habacker)

• Really build x86_64 Windows binaries in Gitlab-CI, instead of building
  i686 binaries a second time (Ralf Habacker)

• When tests will be run using Wine, use STABS debug symbol format so
  that Wine can display backtraces (dbus#133, dbus!104; Ralf Habacker)

dbus 1.13.14 (2020-04-21)
=========================

The “mystery allium” release.

Dependencies:

• On Unix platforms, if getpwnam_r() and getgrnam_r() are implemented,
  they must be POSIX-conformant. The non-POSIX signature seen in ancient
  Solaris versions will no longer work. (dbus!11, Simon McVittie)

Enhancements:

• D-Bus Specification 0.36:
  · Fix a typo in an annotated hexdump of part of a message
    (dbus!152, Zygmunt Krynicki)

• On Linux, use getrandom(2) in preference to /dev/urandom
  (dbus!147, Natanael Copa)

• Add a --sender option to dbus-send, which requests a name and holds it
  until the signal has been sent. (dbus!116, Christopher Morin)

Fixes:

• Fix a crash when the dbus-daemon is terminated while one or more
  monitors are active (dbus#291, dbus!140; Simon McVittie)

• Fix several test failures if the build-time tests were run as uid 0.
  Note that running the tests with elevated privileges is likely to be
  insecure, and should only be attempted in an expendable container or
  virtual machine. (dbus!117, Simon McVittie)

• Fix an assertion failure if a client encounters an out-of-memory
  condition while sending its response to the "OK" authentication
  message, and processing of the "OK" message is subsequently retried
  when more memory is available (dbus!119, Simon McVittie)

• Don't leak struct addrinfo if we run out of memory during a TCP
  connect()
  (dbus!143, dbus!144, Coverity CID 354880; Ralf Habacker, Simon McVittie)

• On Linux with SELinux, don't assume that the system policy has the
  "dbus" security class or the associated AV
  (dbus#198, dbus!128; Laurent Bigonville)

• Handle dbus_connection_set_change_sigpipe() in a thread-safe way
  (dbus!132; Simon McVittie, Ralf Habacker)

• On Unix, use POSIX <poll.h> in preference to <sys/poll.h>
  (dbus!148, Natanael Copa)

• When building with CMake, cope with libX11 in a non-standard location
  (dbus!129, Tuomo Rinne)

• On Windows with verbose mode enabled and outputting to the debug port,
  use a dynamically-allocated buffer to avoid potential stack buffer
  overflows in long messages (dbus#45, dbus!133; Ralf Habacker)

• The dbus-send(1) man page now documents --bus and --peer instead of
  the old --address synonym for --peer, which has been deprecated since
  the introduction of --bus and --peer in 1.7.6
  (fd.o #48816, dbus!115; Chris Morin)

• Fix a wrong environment variable name in dbus-daemon(1)
  (dbus#275, dbus!122; Mubin, Philip Withnall)

• Fix formatting of dbus_message_append_args example
  (dbus!126, Felipe Franciosi)


vdr-plugin-epgsearch/createcats.c
/*                                                                  -*- c++ -*-
Copyright (C) 2004-2013 Christian Wieninger

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
Or, point your browser to http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

The author can be reached at cwieninger@gmx.de

The project's page is at http://winni.vdr-developer.org/epgsearch
*/

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <getopt.h>

#define MINAPPEARANCE 100 // the minimum appearance of a category
#define MAXVALUES 60 // the maximum of values for a category
#define MAXNAMELENGTH 30 // the maximum length of a category name or value


// some helping stuff copied from VDR sources
#define KILOBYTE(n) ((n) * 1024)
#define MAXPARSEBUFFER KILOBYTE(10)

#ifdef __FreeBSD__
#ifdef isnumber
#undef isnumber
#endif
#endif

bool isnumber(const char *s)
{
    if (!*s)
        return false;
    while (*s) {
        if (!isdigit(*s))
            return false;
        s++;
    }
    return true;
}

// --- cReadLine -------------------------------------------------------------
class cReadLine
{
private:
    char buffer[MAXPARSEBUFFER];
public:
    cReadLine() {
        buffer[0] = 0;
    }
    char *Read(FILE *f);
};

char *cReadLine::Read(FILE *f)
{
    if (fgets(buffer, sizeof(buffer), f) > 0) {
        int l = strlen(buffer) - 1;
        if (l >= 0 && buffer[l] == '\n')
            buffer[l] = 0;
        return buffer;
    }
    return NULL;
}

char *skipspace(const char *s)
{
    while (*s && isspace(*s))
        s++;
    return (char *)s;
}

int comparevalue(const void *arg1, const void *arg2)
{
    char* value1 = *(char**) arg1;
    char* value2 = *(char**) arg2;
    return strcmp(value1, value2);
}

// --- cCat -------------------------------------------------------------
class cCat
{
public:
    int appeared;
    char name[MAXPARSEBUFFER];
    int numvalues;
    char** values;

    cCat(char* n)
        : appeared(0), numvalues(0), values(NULL) {
        strcpy(name, n);
    }
    void addvalue(char* value) {
        if (valueexists(value))
            return;
        char* newvalue = (char*) malloc(sizeof(char) * (strlen(value) + 1));
        strcpy(newvalue, value);
        char **tmp = (char**) realloc(values, sizeof(char*) * (numvalues + 1));
        if (tmp) {
            values = tmp;
            values[numvalues++] = newvalue;
        } else {
            free(newvalue);
        }
    }
    bool valueexists(char* value) {
        for (int i = 0; i < numvalues; i++)
            if (strcmp(values[i], value) == 0)
                return true;
        return false;
    }
    void sort() {
        qsort(values, numvalues, sizeof(char*), comparevalue);
    }
};

int comparecat(const void *arg1, const void *arg2)
{
    cCat* cat1 = *(cCat**) arg1;
    cCat* cat2 = *(cCat**) arg2;
    if (cat1->appeared == cat2->appeared) return 0;
    if (cat1->appeared < cat2->appeared) return 1;
    else return -1;
}

// --- cCats -------------------------------------------------------------
class cCats
{
private:
    int numcats;
    cCat** cats;
public:
    cCats(): numcats(0), cats(NULL) {}

    int num() {
        return numcats;
    }

    cCat* add(char* name) {
        cCat* newCat = new cCat(name);
        cCat **tmp = (cCat**) realloc(cats, sizeof(cCat*) * (numcats + 1));
        if (tmp) {
            cats = tmp;
            cats[numcats++] = newCat;
            return newCat;
        } else {
            delete newCat;
            return NULL;
        }
    }

    cCat* get(int i) {
        if (i >= 0 && i < numcats)
            return cats[i];
        else
            return NULL;
    }

    cCat* exists(char* name) {
        for (int i = 0; i < numcats; i++)
            if (strcmp(cats[i]->name, name) == 0)
                return cats[i];
        return NULL;
    }
    void sort() {
        for (int i = 0; i < numcats; i++)
            cats[i]->sort();
        qsort(cats, numcats, sizeof(cCat*), comparecat);
    }
};

int main(int argc, char *argv[])
{
    FILE* f = NULL;
    cCats catlist;
    unsigned int minappearance = MINAPPEARANCE;
    unsigned int maxvalues = MAXVALUES;
    unsigned int maxlength = MAXNAMELENGTH;

    static const struct option long_options[] = {
        { "minappearance", required_argument, NULL, 'm' },
        { "maxvalues",   required_argument, NULL, 'v' },
        { "maxlength",   required_argument,       NULL, 'l' },
        { "help",     no_argument,       NULL, 'h' },
        { NULL,     no_argument,       NULL, 0 }
    };

    int c;
    while ((c = getopt_long(argc, argv, "m:v:l:h", long_options, NULL)) != -1) {
        switch (c) {
        case 'm':
            if (isnumber(optarg)) {
                minappearance = atoi(optarg);
                break;
            }
            fprintf(stderr, "invalid parameter minappearance: %s\n", optarg);
            return 2;
            break;
        case 'v':
            if (isnumber(optarg)) {
                maxvalues = atoi(optarg);
                break;
            }
            fprintf(stderr, "invalid parameter maxvalues: %s\n", optarg);
            return 2;
            break;
        case 'l':
            if (isnumber(optarg)) {
                maxlength = atoi(optarg);
                break;
            }
            fprintf(stderr, "invalid parameter maxlength: %s\n", optarg);
            return 2;
            break;
        case 'h':
            printf("usage: createcats [OPTIONS] /path_to/epg.data\n\n");
            printf("-m N, --minappearance=N    the minimum number a category has to appear\n");
            printf("                           to be used\n");
            printf("-v N, --maxvalues=N        values of a category are omitted if they exceed\n");
            printf("                           this number\n");
            printf("-l N, --maxlength=N        the maximum length of a text to be accepted\n");
            printf("                           as a category value\n");
            printf("-h, --help                 this help\n\n");
            return 0;
        default:
            break;
        }
    }

    if (argc < 2) {
        fprintf(stderr, "ERROR: please pass your epg.data\nusage: createcats epg.data\n");
        return 1;
    }

    f = fopen(argv[argc - 1], "r");
    if (f == NULL) {
        fprintf(stderr, "ERROR: could not open: %s\n", argv[1]);
        return 1;
    }

    char *s;
    cReadLine ReadLine;
    while ((s = ReadLine.Read(f)) != NULL) {
        if (*s == 'D') {
            s = strchr(s, '|'); // jump to possibly first category
            if (!s)
                continue;
            s++;
            char *pstrSearchToken;
            char *pstrSearch = strdup(s);
            pstrSearchToken = strtok(pstrSearch, "|");

            while (pstrSearchToken) {
                // must have a ':'
                char* szPos = NULL;
                if ((szPos = strchr(pstrSearchToken, ':')) == NULL) {
                    pstrSearchToken = strtok(NULL, "|");
                    continue;
                }

                char catname[MAXPARSEBUFFER] = "";
                char catvalue[MAXPARSEBUFFER] = "";

                strncpy(catname, pstrSearchToken, szPos - pstrSearchToken);
                catname[szPos - pstrSearchToken] = 0;
                strcpy(catvalue, skipspace(szPos + 1));

                cCat* cat = catlist.exists(catname);
                if (!cat && strlen(catname) < maxlength) // accept only names up to 30 chars
                    cat = catlist.add(catname);

                if (cat) {
                    cat->appeared++;
                    if (strlen(catvalue) < maxlength) // accept only values up to 30 chars
                        cat->addvalue(catvalue);
                }

                pstrSearchToken = strtok(NULL, "|");
            }
            free(pstrSearch);
        }
    }
    fclose(f);

    catlist.sort();

    f = fopen("epgsearchcats.conf", "w");
    if (f == NULL) {
        fprintf(stderr, "ERROR: could not open outputfile\n");
        return 1;
    }
    fprintf(f, "# -----------------------------------------------------------------------------\n");
    fprintf(f, "# This is just a template based on your current epg.data. Please edit!\n");
    fprintf(f, "# Perhaps a category or its value list should be removed. Also the\n");
    fprintf(f, "# 'name in menu' should be adjusted to your language.\n");
    fprintf(f, "# The order of items determines the order listed in epgsearch. It does not\n");
    fprintf(f, "# depend on the ID, which is used by epgsearch.\n");
    fprintf(f, "# Format:\n");
    fprintf(f, "# ID|category name|name in menu|values separated by ',' (option)|searchmode(option)\n");
    fprintf(f, "# - 'ID' should be a unique positive integer\n");
    fprintf(f, "#   (changing the id later on will force you to reedit your search timers!)\n");
    fprintf(f, "# - 'category name' is the name in your epg.data\n");
    fprintf(f, "# - 'name in menu' is the name displayed in epgsearch.\n");
    fprintf(f, "# - 'values' is an optional list of possible values\n");
    fprintf(f, "#   if you omit the list, the entry turns to an edit field in epgsearch,\n");
    fprintf(f, "#   else it's a list of items to select from\n");
    fprintf(f, "# - 'searchmode' is an optional parameter specifying the mode of search:\n");
    fprintf(f, "#     text comparison:\n");
    fprintf(f, "#     0 - the whole term must appear as substring\n");
    fprintf(f, "#     1 - all single words (delimiters are ',', ';', '|' or '~')\n");
    fprintf(f, "#         must exist as substrings. This is the default search mode.\n");
    fprintf(f, "#     2 - at least one word (delimiters are ',', ';', '|' or '~')\n");
    fprintf(f, "#         must exist as substring.\n");
    fprintf(f, "#     3 - matches exactly\n");
    fprintf(f, "#     4 - regular expression\n");
    fprintf(f, "#     numerical comparison:\n");
    fprintf(f, "#     10 - less\n");
    fprintf(f, "#     11 - less or equal\n");
    fprintf(f, "#     12 - greater\n");
    fprintf(f, "#     13 - greater or equal\n");
    fprintf(f, "#     14 - equal\n");
    fprintf(f, "#     15 - not equal\n");
    fprintf(f, "# -----------------------------------------------------------------------------\n\n");
    int id = 1;
    for (int i = 0; i < catlist.num(); i++) {
        cCat* cat = catlist.get(i);
        if (cat->appeared > (int)minappearance && cat->numvalues > 1) { // accept only category, that have at least 2 values and appear more than MINAPPEARANCE timers
            fprintf(f, "# '%s' found %d times with %d different values %s\n", cat->name, cat->appeared, cat->numvalues, cat->numvalues >= (int)maxvalues ? "(values omitted, too much)" : "");
            fprintf(f, "%d|%s|%s|", id++, cat->name, cat->name);
            for (int j = 0; cat->numvalues < (int)maxvalues && j < cat->numvalues; j++)
                fprintf(f, "%s%s", cat->values[j], (j == cat->numvalues - 1 ? "" : ","));
            fprintf(f, "|1\n\n");
        }
    }
    fclose(f);

    printf("epgsearchcats.conf created!\n");
    return 0;
}
